### ISSUE TITLE
Unable to Merge contacts?

### ISSUE THREAD

**Issue by davidjosephhayes** [twomice/com.joineryhq.activityical#38]
I think a unique contraint on the `civicrm_activityicalcontact` table prevents contact from being merged. However, could be something else; I do find it odd I would be the first to run across this.

# Error Log
```
Mar 12 15:58:19  [error] $Fatal Error Details = Array
(
    [callback] => Array
        (
            [0] => CRM_Core_Error
            [1] => handle
        )

    [code] => -5
    [message] => DB Error: already exists
    [mode] => 16
    [debug_info] => UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 574 [nativecode=1062 ** Duplicate entry '106' for key 'contact_id']
    [type] => DB_Error
    [user_info] => UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 574 [nativecode=1062 ** Duplicate entry '106' for key 'contact_id']
    [to_string] => [db_error: message="DB Error: already exists" code=-5 mode=callback callback=CRM_Core_Error::handle prefix="" info="UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 574 [nativecode=1062 ** Duplicate entry '106' for key 'contact_id']"]
)


Mar 12 15:58:19  [debug] $backTrace = #0 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Core/Error.php(205): CRM_Core_Error::backtrace("backTrace", TRUE)
#1 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php(922): CRM_Core_Error::handle(Object(DB_Error))
#2 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/vendor/pear/db/DB.php(997): PEAR_Error->__construct("DB Error: already exists", -5, 16, (Array:2), "UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 57...")
#3 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php(575): DB_Error->__construct(-5, 16, (Array:2), "UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 57...")
#4 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php(223): PEAR->_raiseError(Object(DB_mysqli), NULL, -5, 16, (Array:2), "UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 57...", "DB_Error", TRUE)
#5 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/vendor/pear/db/DB/common.php(1928): PEAR->__call("raiseError", (Array:7))
#6 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/vendor/pear/db/DB/mysqli.php(936): DB_common->raiseError(-5, NULL, NULL, "UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 57...", "1062 ** Duplicate entry '106' for key 'contact_id'")
#7 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/vendor/pear/db/DB/mysqli.php(406): DB_mysqli->mysqliRaiseError()
#8 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/vendor/pear/db/DB/common.php(1234): DB_mysqli->simpleQuery("UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 574")
#9 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/packages/DB/DataObject.php(2696): DB_common->query("UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 574")
#10 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/packages/DB/DataObject.php(1829): DB_DataObject->_query("UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 574")
#11 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Core/DAO.php(445): DB_DataObject->query("UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 574")
#12 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Core/DAO.php(1551): CRM_Core_DAO->query("UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 574", TRUE)
#13 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Dedupe/Merger.php(601): CRM_Core_DAO::executeQuery("UPDATE civicrm_activityicalcontact SET contact_id = 106 WHERE contact_id = 574", (Array:0), TRUE, NULL, TRUE)
#14 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Dedupe/Merger.php(1415): CRM_Dedupe_Merger::moveContactBelongings(Object(CRM_Dedupe_MergeHandler), (Array:3), (Array:0))
#15 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Contact/Form/Merge.php(308): CRM_Dedupe_Merger::moveAllBelongings(106, 574, (Array:11))
#16 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Core/Form.php(507): CRM_Contact_Form_Merge->postProcess()
#17 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Core/StateMachine.php(144): CRM_Core_Form->mainProcess()
#18 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Core/QuickForm/Action/Next.php(43): CRM_Core_StateMachine->perform(Object(CRM_Contact_Form_Merge), "next", "Next")
#19 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/packages/HTML/QuickForm/Controller.php(203): CRM_Core_QuickForm_Action_Next->perform(Object(CRM_Contact_Form_Merge), "next")
#20 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/packages/HTML/QuickForm/Page.php(103): HTML_QuickForm_Controller->handle(Object(CRM_Contact_Form_Merge), "next")
#21 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Core/Controller.php(347): HTML_QuickForm_Page->handle("next")
#22 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Utils/Wrapper.php(98): CRM_Core_Controller->run()
#23 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php(285): CRM_Utils_Wrapper->run("CRM_Contact_Form_Merge", "Merge Contact", (Array:0))
#24 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php(68): CRM_Core_Invoke::runItem((Array:13))
#25 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php(36): CRM_Core_Invoke::_invoke((Array:3))
#26 /home/site.com/public_html/wp-content/plugins/civicrm/civicrm.php(1644): CRM_Core_Invoke::invoke((Array:3))
#27 /home/site.com/public_html/wp-includes/class-wp-hook.php(292): CiviCRM_For_WordPress->invoke("")
#28 /home/site.com/public_html/wp-includes/class-wp-hook.php(316): WP_Hook->apply_filters("", (Array:1))
#29 /home/site.com/public_html/wp-includes/plugin.php(484): WP_Hook->do_action((Array:1))
#30 /home/site.com/public_html/wp-admin/admin.php(259): do_action("toplevel_page_CiviCRM")
#31 {main}
```

# Possible Solution
Add this hook:
```
function activityical_civicrm_merge($type, &$data, $mainId = NULL, $otherId = NULL, $tables = NULL) {

	// remove activityical table in merge automatically added from foreign key
	if ($type==='cidRefs' && !empty($data['civicrm_activityicalcontact']))
		unset($data['civicrm_activityicalcontact']);

}
```

**Comment by twomice** [comment]
Noted, thanks for the report @davidjosephhayes We'll try and repro here and get back with you.

**Comment by lucky091588** [comment]
Hi @davidjosephhayes, I tried reproducing the error but failed. 

Can you provide specific steps to produce the error and what CiviCRM version are you using?

Thanks

**Comment by davidjosephhayes** [comment]
Ya, I'll setup a dev site and reproduce.

**Comment by lucky091588** [comment]
Thanks @davidjosephhayes, I'll watch for that update.

**Comment by lucky091588** [comment]
Hi @davidjosephhayes, any progress on this issue?
Can you provide specific steps to produce the error and what CiviCRM version are you using?

Thanks

**Comment by davidjosephhayes** [comment]
Apologies for the delay.

I was able to reproduct this on a new CiviCRM install. Here are the steps I took:

1. Install CiviCRM
2. Install CiviCRM Activity iCalendar Feed
3. Create a group "Activity iCalendar"
![image](https://user-images.githubusercontent.com/9373380/120537152-cf48cf00-c399-11eb-898d-de3752d2301c.png)
4. Create two contacts
![image](https://user-images.githubusercontent.com/9373380/120537088-b9d3a500-c399-11eb-8f3b-5b98082b14b8.png)
5. Add these to contacts to the "Activity iCalendar" Group
![image](https://user-images.githubusercontent.com/9373380/120537024-a7f20200-c399-11eb-8695-abff315b6acf.png)
6. View the feel for each contact to trigger a record creation in the `civicrm_activityicalcontact` table
![image](https://user-images.githubusercontent.com/9373380/120536842-6c573800-c399-11eb-8055-f57f47e9afda.png)
7. Attempt to merge the two contacts
![image](https://user-images.githubusercontent.com/9373380/120537258-edaeca80-c399-11eb-8cd7-8ced22227469.png)
8. Receive `DB Error: already exists` error message
![image](https://user-images.githubusercontent.com/9373380/120537319-015a3100-c39a-11eb-9eb2-a0801638432f.png)

I have a dev site setup currently that will show this error. Happy to share any credential that would be helpful.

**Comment by lucky091588** [comment]
Thanks @davidjosephhayes. I will try to reproduce the error using these steps.

**Comment by twomice** [comment]
Fixed in #43, slightly different approach to OP. Thanks to @davidjosephhayes for the report and the initial concept.

**Comment by davidjosephhayes** [comment]
Very cool!
